nova/scheduler/weights/affinity.py

```diff
@@ -70,6 +77,36 @@ def weight_multiplier(self):
 
         return CONF.filter_scheduler.soft_affinity_weight_multiplier
 
+    def _weigh_object(self, host_state, request_spec):
+        weight = super(ServerGroupSoftAffinityWeigher, self)._weigh_object(
+            host_state, request_spec)
+
+        # if the host contained servers from the same group, return that weight
+        if weight:
+            return weight
+
+        # get shards of the host
+        host_shard_aggr = [aggr for aggr in host_state.aggregates
+                            if aggr.name.startswith(self._SHARD_PREFIX)]
+        if len(host_shard_aggr) > 1:
+            LOG.warning('More than one host aggregates found for '
+                        'host %(host)s, selecting first.',
+                        {'host': host_state.host})
+            host_shard_aggr = [host_shard_aggr[0], ]
```

Comment:
```python
host_shard_aggrs = …
if not host_shards_arggs:
    LOG.warning(…
    return 0
if len(host_shard_aggrs) > 1:
    LOG.warning(…
host_shard_aggr = host_shard_aggrs[0]
```
That's how I envision it. No need to carry around a list if we just have a single element.